V3.3 renesas dom0

firscity opened the pull request:
Remastered and cleaned Xentroops Dom0 changes, applied over zephyr-v3.3.0 branch of Zephyr OS.

firscity committed:
xenvm: dom0: update Xen public headers for running Zephyr as Dom0

Xen public headers are added to Zephyr kernel partially, so we need
to update them for new features implementation.

Further implementation of Xen domains configuration and memory
management requires dedicated Xen public headers inside Zephyr kernel.
It will be used for add Zephyr OS Xen Domain-0 funtionalities.

Update existent and add new required public headers from Xen 4.17.0
release.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xen: change HVM functions signature to run it for other domains

This commit adds possibility to call hypervisor HVM parameter functions
for specified domain (instead of only DOMID_SELF). It is needed for
configuring domains, that were created from Zephyr control domain.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
boards: arm64: add Renesas H3ULCB CA57 board support

Add basic functionality for supporting h3ulcb board:
  * add documentation for h3ulcb board;
  * add pinctrl dtsi for scif driver;
  * add dts, yaml and configuration files.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
soc: arm64: add Renesas Rcar Gen3 SoC support

Added files for supporting  arm64 Renesas r8a77951 SoC.

Enable build of clock_control_r8a7795_cpg_mssr.c for
a new ARM64 SoC R8A77951.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

firscity committed:
xenvm: events: add event channel allocation for domain-0

This commit adds interface for evtchn allocation that can be used by
privilaged domain. Domain 0 can specify both dom and remote_dom
parameters for hypercall, where in others domains dom should be always
DOMID_SELF. It is needed for creating pre-defined channels during
domain setup in Zephyr Dom0.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: sdhc: add support of Renesas MMC driver

Add basic functionality of Renesas SD/MMC driver. It can be used
for both gen3 and gen4 R-car boards, but tested only with H3ULCB,
Salvator XS M3 and Spider boards. This driver working with SDHC
subsystem.

The driver supports regularal reading/writing throught SD/MMC
controller buffer, DMA mode w/o interrupts and timing tuning.

Notes:
    * the driver doesn't support SPI mode;
    * SCC tuning and DMA mode based on IRQs are enabled by default;
    * an address of a data buffer has to be aligned to 128 bytes if it
      is not, driver will use non-DMA mode automatically;
    * Renesas MMC DMAC doesn't support 64-bit DMA addresses, so for
      case when we have 64-bit xref data address we use non-DMA mode;
    * SD/MMC controller supports block size between 512 and 1 with
      a lot of restrictions, more details you can find in code;
    * support of HS400 mode isn't implemented inside driver;
    * tuning of SD card modes hasn't been tested yet.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

dsemenets committed:
subsys/storage/flash_map: remove unnecessary includes

Those files includes soc.h header which exists not for all boards.
soc.h consists soc-depended defenitions and need to be included by
soc-depended sources

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
soc: arm64: xenvm: move XEN_INITIAL_DOMAIN config to Xen driver

Untie XEN_INITIAL_DOMAIN from XenVM SoC configuration, it is
necessary for running Xen dom-0 on different boards. Move it from
XenVM SoC configuration to Xen driver. Set XEN_INITIAL_DOMAIN by
default to n in order to prevent issues with existing setups,
which may use XenVM as DomU.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: uart_rcar: do MMIO mapping inside driver

Add MMIO mapping for UART Renesas driver in order to avoid
mappings inside mmu_regions.c file(s).

There are a lot of changes inside SCIF driver inside this
commit, because reg addr and size may be stored in RAM or
ROM and appropriately in different driver structures data
or cfg, and, because the name of field reg base is changed.

Note: it is common approach according to Zephyr documentation.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: pfc_rcar: do MMIO mapping inside driver

Add MMIO mapping for PFC Renesas driver in order to avoid
mappings inside mmu_regions.c file.

Add a new system init function pfc_rcar_driver_init to PFC
Renesas driver for invoking a memory mapping macro.
Note: PFC Renesas driver doesn't use Zephyr Device Model.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
dts: bindings: xen: add xen,xen.yaml file

Add yaml file for 'xen,xen', because without it an appropriate
'CONFIG_DT_HAS_XEN_XEN_ENABLED' isn't generated.

Remove xen,xen-4.15.yaml at all, because it isn't necessary to have
yaml for some specific Xen version.

Add subdirectory of 'xen' when we have Xen Hypervisor node enabled
in dts. It is necessary action if we want to use driver for Xen
Initial domain shield, that means that the driver isn't tied only
to XenVM board configuration.

Add subdirectory of 'xen' when we have Xen Hypervisor node enabled
in dts. It is necessary action if we want to use arch arm64 core Xen
related code for Xen Initial domain shield, that means that the core
of Xen isn't tied only to XenVM SoC configuration.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
shields: xen_dom0: add overlay for h3ulcb board

Delete ram and redefine memory node for h3ulcb board, because
reg address and size can be different for bar-metal and Xen
dom0 builds;

Delete scif node for h3ulcb board, because dom0 uses Xen
Hypervisor takes serial and uses own node for this.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
boards: shields: add salvator xs m3 overlay to xen dom0 shield

Delete ram and redefine memory node for Salvator XS M3 board,
because reg address and size can be different for bare-metal
and Xen dom0 builds.

Delete scif node for Salvator XS M3 board, because dom0 uses
Xen Hypervisor takes serial and uses own node for this.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
soc: arm64: add support of r8a779f

Add support of r8a779f SoC and appropriate dts file
with minimal set of nodes.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
boards: shields: add Renesas Spider overlay to Xen dom0 shield

Delete ram and redefine memory node for Renesas Spider board,
because reg address and size can be different for bare-metal
and Xen dom0 builds.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
fs: littlefs: remove dependency on flash map from FILE_SYSTEM_LITTLEFS

LittleFS may be used not only for Flash devices, we can use it for block
devices too. So, I have changed dependency rules and remove dependency
on flash map from FILE_SYSTEM_LITTLEFS.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>
Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
fs: littlefs: add ability to build little fs without flash map

LittleFS may be used not only for Flash devices, we can use it for
block devices too.

Add possibility of building LittleFS for case when
CONFIG_FLASH_MAP and CONFIG_FLASH_PAGE_LAYOUT are disabled.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
samples: littlefs: remove flash map configs when we use block device

We don't need anymore enabling of Flash map for building littlefs
sample with block disk device config.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
arch: arm64: mmu: WA: avoid using of set/way cache instructions for Xen

Unfortunately, set/way ops are not easily virtualized by Xen. S/W
emulation is disabled, because IP-MMU is active for Dom0. IP-MMU is
a IO-MMU made by Renesas, as any good IO-MMU, it shares page-tables
with CPU. Trying to emulate S/W with IP-MMU active will lead to
IO-MMU faults. So if we build Zephyr as a Xen Initial domain, it
won't work with cache management support enabled.

In this commit I use VA data invalidate instead of set/way --
invalidate all virtual address space, so it might increase system
initialization time.

Note: this patch is applied as a simple and fast workaround, we need
      to find more optimal solution.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: add generic Renesas functions for get/set rate

Add associated tables of clocks and API for working with these tables,
from this moment the relationship between clocks and their divider are
built. After set rate of some Core clock, driver has to update all
in/out rates of all childrens recursively. During get/set rate calls
if out rate is unknown, we try to get parent in/out rates and its
divider, in case when parent doesn't have valid in/out rates we get
parent of parent and so on until we get parent with a valid in or out
rates.

Add generic Renesas functions for get/set rate of CPG.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: redesign cpg driver for r8a7795

Add associative tables for core and mod clocks.

Add possibility to enable/disable any of core clocks which presents
in the associative table.

Add handler for setting rate to driver and use generic rcar cpg
function for getting rate.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: delete irq lock/unlock from mstp_clock_endisable

Delete IRQ lock/unlock calls from 'rcar_cpg_mstp_clock_endisable',
because 'rcar_cpg_mstp_clock_endisable' function is always called
under spin lock.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: do MMIO mapping inside driver

Add MMIO mapping for Renesas CPG driver in order to avoid
mappings inside mmu_regions.c file. Remove MMU region for
Renesas CPG driver.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: xen: gnttab: do Xen node mapping inside driver

Move memory mapping of Xen node to Grant Table driver system init
function. After moving mapping we don't need anymore records of
xen-xen node into 'mmu_regions' array, so they were deleted from
all SoCs: Rcar Gen3/Gen4 and XenVM.

We need at least 16M of virtual address space to map memory of Xen
node, so the virtual memory sized has been increased to 32 MB, it
should be enough for basic use-cases and mapping of 16M mem region
of Xen node.

Unfortunately, after moving we also need to increase number of XLAT
tables. The previous code was more efficient if we talking about
usage of XLAT tables, because it mapped grant tables using a higher-
order table that allows mapping blocks of 2MB. And after the changes
is maps every 4KB page, so we need more XLAT tables.

Note: Xen Grant Table driver doesn't use Zephyr Device Model.

Authored-by: Mykola Kvach <mykola_kvach@epam.com>
Co-authored-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: pinctrl: pfc_rcar: add support of voltage control to pfc driver

Add support of voltage control to Renesas PFC driver. Voltage register
mappings have been added to r8a7795, r8a7796 and r8a779f SoCs.

Allow 'power-source' property for 'renesas,rcar-pfc' node. This property
will be used for configuring IO voltage on appropriate pin. For now it
is possible to have only two voltages: 1.8 and 3.3.

Note: it is possible to change voltage only for SD/MMC pins on r8a7795
      and r8a7796 SoCs, r8a779f supports more pins for configuring
      voltage, but only SD/MMC pins have been added to the array.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
sd: request signal voltage according to host capabilities

The SD initialization function (sd_init_io) may be also used for
initialization of MMC/eMMC. eMMC/MMC may not allow 3.3 signal voltage
by default, so we should check which voltage during initialization
we may use.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: move cpg log module declare to common rcar cpg file

Move cpg log module declare to common rcar cpg file. If the module
consists of multiple files, then LOG_MODULE_REGISTER should appear
in exactly one of them. Each other file should use LOG_MODULE_DECLARE
to declare its membership in the module.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

firscity committed:
xen: implement Xen domain control interface

Add Xen domctl API implementation for Zephyr as control domain.

Previously Zephyr OS was used as unprivileged Xen domain (Domain-U),
but it also can be used as lightweight Xen control domain (Domain-0).
To implement such fuctionality additional Xen interfaces are needed.
One of them is Xen domain controls (domctls) - it allows to create,
configure and manage Xen domains.

Also, used it as a possibility to update files copyright and licenses
identifiers in touched files.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xen: add helper functions for Xen domain memory management

Add wrapper functions for Xen memory managment hypercall. They can be
used for unprivilaged Zephyr guest initialization and for domain
management, when Zephyr is used as Xen Domain 0.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
soc: arm64: add support of r8a77961

Add support of r8a77961 SoC to gen3 series.
Create a dtsi file with a common part for both r8a77951 and r8a77961.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
soc: arm64: add PFC files to Renesas r8a77961 Gen3 SoC

Add Pin Function Controller tables of registers and their bits
for ARM64 Renesas R-Car family. With this changes we can use
Renesas PFC driver for configuring bias and driving capabilities.

Add only needed driver strength and bias pins to PFC,
e.g. SDx and UART TX/RX pins.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
boards: arm64: add support of Salvator XS M3 board

Add support of 'rcar_salvator_xs_m3' board: minimal dts
and configuration.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
soc: arm64: add PFC files to Renesas r8a77951 Gen3 SoC

Add Pin Function Controller tables of registers and their bits
for ARM64 Renesas R-Car family. With this changes we can use
Renesas PFC driver for configuring bias and driving capabilities.

Note: some of files copy-pasted from Renesas Arm 32
      SoC directory and this commit is a temporary
      solution, we need to rework it in order to use
      the same source files for both architectures.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
boards: arm64: add support of Renesas Spider board

Add support of 'rcar_spider' board: minimal dts
and configuration.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: clock_control: add support of r8a779f cpg driver

Add support of r8a779f cpg driver

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
soc: arm64: add support of pinctrl for r8a779f

Add support of pinctrl for r8a779f.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
tests: add h3ulcb, salvator xs m3 and spider configs to perf disk tests

Add MMC related cfgs/overlays of arm64 gen3/gen4 rcar boards to perf and
susys/sd/mmc tests.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: regulator-fixed: extend api of driver (set/list/count voltages)

Allow to create fixed-regulator nodes without gpio for en/dis power
source. Allow properties 'regulator-min-microvolt' and
'regulator-max-microvolt' for fixed refulator. Note: they should be
equal. Add simple functions for getting list of allowed voltages, count
of voltages and dummy function for set voltage, it is used only for
checking voltage range inside regulator_set_voltage function.

The fake set voltage function inside regulator-fixed is needed only
to have common interface with other regulators and for avoiding extra
checks in place where 'regulator_set_voltage' is invoked.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
boards: arm64: rcar: add board-specific configuration for MMC node

Add 'power-source' property equaled to 1.8V to every eMMC pin in
order to use this config inside Renesas SD/MMC driver.

Add pin control group for UHS modes for H3ULCB, Salvator XS M3
and Spider boards to appropriate dts files. Both 'uhs' and
'default' pins states have the same properties for eMMC,
e.g. 1.8V.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

firscity committed:
boards: shields: add virtual Xen dom0 shield support

Zephyr board shields allow to extend board features and services. It is
made by applying shield generic Kconfig/device-tree changes and
board-specific device-tree overlays. Originally it was intended to be
used for easier board prototyping and compatibility with applications.

Also it looks like great approach for convertion Zephyr OS build to Xen
control domain on different boards, as it requires Kconfig definition
and board device tree changes (configuration for memory node, hypervisor
console, grant table region etc).

Please note, that Xen hypervisor passes all selected parameters via
domain device-tree, but Zephyr does not use it in runtime. So, user
should keep values in board overlays in sync with real Xen values.

Add virtual Xen dom0 shield support and the dependence of config
UART_XEN_HVC_CONSOLEIO on configuration of shield 'xenvm_dom0', as this
way for console communication may be used as default only by privileged
Xen domain.

Add sample shield and docs for converting qemu_cortex_a53 board to Xen
Domain-0 guest. This approach might be used for other boards, where
Zephyr Dom0 is needed.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
samples: littlefs: add support for H3ULCB, Salvator XS M3 and Spider

Add support of R-Car H3ULCB, Salvator XS M3 and Spider boards to
littlefs sample.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

oleksiimoisieiev committed:
mem_manage: Cast types in assert call to uintptr_t

Cast VM_BASE and VM_SIZE to uintptr_t because compiler treats that
defies as int by default, which could lead us to the compilation
warning when VM_BASE + VM_SIZE > MAX_INT.
In Zephyr uintptr_t is long unsigned int so we should cast VM_BASE +
VM_SIZE to uintptr_t to compare with addr, which is uintptr_t as well.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

xakep-amatop commented:
Hi, @firscity 
please,

* move the last three commits (little fs) before the next commit [samples: littlefs: add support for H3ULCB, Salvator XS M3 and Spider](https://github.com/xen-troops/zephyr/pull/43/commits/82f4ee583f1476f6926c29cc6a0e76e6d0b480b4)
* change commit massage of the commit https://github.com/xen-troops/zephyr/pull/43/commits/b6bf3e56c65f3347757f981d3d60f833a5aebe7a to the next one:
```
drivers: sdhc: add support of Renesas MMC driver

Add basic functionality of Renesas SD/MMC driver. It can be used
for both gen3 and gen4 R-car boards, but tested only with H3ULCB,
Salvator XS M3 and Spider boards. This driver working with SDHC
subsystem.

The driver supports regularal reading/writing throught SD/MMC
controller buffer, DMA mode w/o interrupts and timing tuning.

Notes:
    * the driver doesn't support SPI mode;
    * SCC tuning and DMA mode based on IRQs are enabled by default;
    * an address of a data buffer has to be aligned to 128 bytes if it
      is not, driver will use non-DMA mode automatically;
    * Renesas MMC DMAC doesn't support 64-bit DMA addresses, so for
      case when we have 64-bit xref data address we use non-DMA mode;
    * SD/MMC controller supports block size between 512 and 1 with
      a lot of restrictions, more details you can find in code;
    * support of HS400 mode isn't implemented inside driver;
    * tuning of SD card modes hasn't been tested yet.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
```
* add changes from the next pull https://github.com/xakep-amatop/zephyr-mmc-rcar/pull/2 to the commit https://github.com/xen-troops/zephyr/pull/43/commits/b6bf3e56c65f3347757f981d3d60f833a5aebe7a

firscity committed:
drivers: xen: return hypercall results from notify_evtchn

Xen event channel notification may fail during hypervisor handling,
so it will be good for user to have a possibility to check results.

Previous implementation ignored hypervisor return code, now it will be
passed to caller.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
drivers: xen: export missing evtchn API functions

Previously some functions, that are a part of Xen event channel API
were missing in events.h header (clear_event_channel() was also marked
as 'static').

Add missing calls to events.h and add documentation description for
to let users add it to external drivers.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop commented on include/zephyr/sys/mem_manage.h:
```diff
@@ -141,15 +141,15 @@ static inline uintptr_t z_mem_phys_addr(void *virt)
-			  (CONFIG_SRAM_SIZE * 1024UL))),
+	__ASSERT((addr >= (uintptr_t)CONFIG_SRAM_BASE_ADDRESS) &&
+		 (addr < ((uintptr_t)CONFIG_SRAM_BASE_ADDRESS +
+			  ((uintptr_t)CONFIG_SRAM_SIZE * 1024UL))),
```
(note) we don't need explicit casts here, because expression result will be unsigned long due to implicit cast

xakep-amatop commented on drivers/xen/events.c:
```diff
@@ -92,7 +92,7 @@ int evtchn_set_priority(evtchn_port_t port, uint32_t priority)
 }
 
-void notify_evtchn(evtchn_port_t port)
+int notify_evtchn(evtchn_port_t port)
```
(question) do we really need this commit as far as I see a caller ignores return status?

xakep-amatop reviewed (approved):
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>`

firscity commented:
This PR will be reopened for final review on target branch.
